django/core/validators.py

```diff
@@ -223,12 +222,21 @@ class EmailValidator:
         r"((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))\Z",
         re.IGNORECASE,
     )
+    domain_idn_regex = _lazy_re_compile(
+        r"^" + DomainNameValidator.hostname_re + DomainNameValidator.domain_re + r"\."
+        r"(?!-)"
+        r"(?:[a-z" + DomainNameValidator.ul + "-]{2,63}|xn--[a-z0-9]{1,59})"
+        r"(?<!-)"
+        r"$",
+        re.IGNORECASE,
+    )
```

Comment:
Rather than having two separate regular expressions, one for ASCII domains and one for IDNs, I think we could use a single regex that matches either, like URLValidator and DomainNameValidator do. Then the domain validation would only need a single re.match() call.

What we're looking for is DomainNameValidator's hostname_re + domain_re + tld_re (a hostname followed by zero or more subdomains followed by a top-level domain). And it turns out URLValidator.host_re is exactly that, which kind of makes sense: the "domain" part of an email address is equivalent to the "host" part of an URL.

So EmailValidator should be able to use something like:

```python
    host_re = URLValidator.host_re
    domain_regex = _lazy_re_compile(
        r"^" + host_re + r"\Z"
    )
```

Reply:
Just to clarify, the existing host_re in URLValidator already supports both IDNs and ASCII-only domains, so there should be no need to include separate handling for ASCII domains.